### PR TITLE
Dynamic Partition Elimination (DPE) for Hash Right Joins

### PR DESCRIPTION
**Pull request by 24nishant**
With this PR, we are adding support of Dynamic Partition Elimination (DPE) for right joins in ORCA. The DPE will take place if the following conditions are satisfied, in a join scenario
1. The outer table must be a partition table
2. Join condition should be on the partitioned key of the outer table
3. The distribution column and partition column should be same for the partitioned table

Note:
1. In the final plan right join will be produced if the following condition
(innerRows * innerWidth * confidenceFactor < outerRows * outerWidth)
defined in CXformRightOuterJoin2HashJoin::Transform(CXformContext *pxfctxt, CXformResult *pxfres, CExpression *pexpr)
is satisfied by the tables.

2. With this PR, we shall also be addressing the following FIXME, mentioned in "src/backend/gporca/server/CMakeLists.txt":
   GPDB_12_MERGE_FIXME: re-add the following test to Right Join once DPE is supported
      RightJoinDPS PartTbl-RightOuterHashJoin-DPE-IsNull

### CODE REVIEW

File: src/backend/gporca/server/CMakeLists.txt [hunk 1]
```diff
@@ -45,8 +45,6 @@ endfunction()
 # IN-ArrayCmp
 # NOT-IN-ArrayCmp
 
-# GPDB_12_MERGE_FIXME: re-add the following test to Right Join once DPE is supported
-# RightJoinDPS PartTbl-RightOuterHashJoin-DPE-IsNull
```
Comment (chrishajas) [inline src/backend/gporca/server/CMakeLists.txt:49]: There's also a RightJoinDPS.mdp test file, can add that to the schedule

File: src/test/regress/sql/join.sql [hunk 2]
```diff
@@ -2269,3 +2269,86 @@ reset enable_hashjoin;
 reset enable_nestloop;
 reset enable_seqscan;
 reset enable_bitmapscan;
+
```
Comment (chrishajas) [inline src/test/regress/sql/join.sql:2272]: Let's put these in join_gp.sql instead to minimize future merge conflicts

File: src/test/regress/sql/join.sql [hunk 3]
```diff
@@ -2269,3 +2269,86 @@ reset enable_hashjoin;
...
+drop table if exists foo;
+drop table if exists bar_PT1;
+drop table if exists bar_PT2;
+
+-- Table creation : Normal table
+create table foo (a int, b int);
+insert into foo select i,i from generate_series(1,5)i;
+analyze foo;
+-- Table creation : First Partitioned table with same 'Distribution Column' and 'Partitioning key'
+create table bar_PT1 (a1_PC int, b1 int) partition by range(a1_PC) (start (1) inclusive end (50) every (2));
```
Comment (chrishajas) [inline src/test/regress/sql/join.sql:2297]: Can we use fewer partitions for the test

File: src/backend/gporca/libgpopt/src/operators/CPhysicalRightOuterHashJoin.cpp [hunk 4]
```diff
@@ -120,4 +121,146 @@ CPhysicalRightOuterHashJoin::CreateOptRequests(CMemoryPool *mp)
...
+			true  // fUseConstraints
+		);
+	}
+
+	return pexprPred;
+}
+
+
+CPartitionPropagationSpec *
+CPhysicalRightOuterHashJoin::PppsRequired(
```
Comment (chrishajas) [inline src/backend/gporca/libgpopt/src/operators/CPhysicalRightOuterHashJoin.cpp:152]: Is this the same as `CPhysicalInnerHashJoin::PppsRequired`? If so, can we refactor both to call the same code

File: src/backend/gporca/libgpopt/src/operators/CPhysicalRightOuterHashJoin.cpp [hunk 5]
```diff
@@ -120,4 +121,146 @@ CPhysicalRightOuterHashJoin::CreateOptRequests(CMemoryPool *mp)
...
+			mp, exprhdl, pppsRequired, child_index, pdrgpdpCtxt, ulOptReq);
+	}
+	return pps_result;
+}
+
+CPartitionPropagationSpec *
+CPhysicalRightOuterHashJoin::PppsDerive(CMemoryPool *mp,
+										CExpressionHandle &exprhdl) const
+{
+	CPartitionPropagationSpec *pps_outer = exprhdl.Pdpplan(0)->Ppps();
```
Comment (chrishajas) [inline src/backend/gporca/libgpopt/src/operators/CPhysicalRightOuterHashJoin.cpp:256]: Same here, if it's the same can we call a common method

File: src/test/regress/sql/join.sql (hunk 3 above)
Comment (24nishant) [reply src/test/regress/sql/join.sql]: I have updated the tests, with fewer number of partitions.

File: src/test/regress/sql/join.sql (hunk 2 above)
Comment (24nishant) [reply src/test/regress/sql/join.sql]: I have moved the tests to join_gp.sql

File: src/backend/gporca/server/CMakeLists.txt (hunk 1 above)
Comment (24nishant) [reply src/backend/gporca/server/CMakeLists.txt:49]: On running these old mdp files, failure was observed in ORCA. Based on the test case information given in these mdp file, I ran only the sql commands, the queries ran successfully. Then I created 2 new mdp files for the same tests and added them. With the new mdp files, no test case failure was observed.

Old vs new mdp files:
1. RightJoinDPS.mdp (OLD) ---> RightJoinDPE.mdp(NEW)
2. PartTbl-RightOuterHashJoin-DPE-IsNull.mdp (OLD) --->PartTbl-RightOuterHashJoin-DPE-IsNull-V2.mdp (NEW)

File: src/test/regress/expected/join_gp_optimizer.out [hunk 6]
```diff
@@ -1627,3 +1627,406 @@ on (coalesce(t.id1) = tq_all.id1  and t.id2 = tq_all.id2) ;
...
+       |    |   |   |  7 |     7
+       |    |   |   |  8 |     8
+     1 |  1 | 1 | 1 |  1 |     1
+     5 |  5 | 5 | 5 |  5 |     5
+       |    |   |   |  6 |     6
+       |    |   |   |  9 |     9
+       |    |   |   | 10 |    10
+       |    |   |   | 11 |    11
+(11 rows)
+
```
Comment (DevChattopadhyay) [inline src/test/regress/expected/join_gp_optimizer.out:2028]: Can you please add some test cases of conjunction and disjunction

File: src/backend/gporca/data/dxl/minidump/PartTbl-RightOuterHashJoin-DPE-IsNull-V2.mdp [hunk 7]
```diff
@@ -0,0 +1,552 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (chrishajas) [inline src/backend/gporca/data/dxl/minidump/PartTbl-RightOuterHashJoin-DPE-IsNull-V2.mdp:1]: add query and intent to these mdps (whether it should generate a partition selector or not)

File: src/test/regress/sql/join_gp.sql [hunk 8]
```diff
@@ -768,3 +768,89 @@ full join ( select r.id1, r.id2 from t_issue_10315 r group by r.id1, r.id2 ) tq_
...
+-- Outer table: Partitioned table, Join Condition on Partition key: Yes, Result: DPE - No
+explain (costs off) select * from bar_PT2 right join foo on bar_PT2.b2_PC =foo.a;
+select * from bar_PT2 right join foo on bar_PT2.b2_PC =foo.a;
+-- Outer,Inner table: Partitioned table, Join Condition on Partition key: Yes, Result: DPE - No
+explain (costs off) select * from bar_PT2 right join bar_PT1 on bar_PT2.b2_PC =bar_PT1.b1;
+select * from bar_PT2 right join bar_PT1 on bar_PT2.b2_PC =bar_PT1.b1;
+
+-- Case 3 : Other Cases
+
+explain (costs off) select * from bar_PT1 right join foo on bar_PT1.a1_PC in (select foo.a from foo where foo.a<3);
```
Comment (chrishajas) [inline src/test/regress/sql/join_gp.sql:834]: Can you add the intent of these cases too. We should have a justification of what we're looking for in each test

File: src/test/regress/sql/join_gp.sql [hunk 9]
```diff
@@ -768,3 +768,89 @@ full join ( select r.id1, r.id2 from t_issue_10315 r group by r.id1, r.id2 ) tq_
...
+select * from bar_PT1 right join foo on bar_PT1.a1_PC in (select foo.a from foo where foo.a<3);
+
+explain (costs off) select * from bar_PT1 right join foo on bar_PT1.a1_PC in (select foo.a from foo where foo.a=bar_PT1.a1_PC);
+select * from bar_PT1 right join foo on bar_PT1.a1_PC in (select foo.a from foo where foo.a=bar_PT1.a1_PC);
+
+-- Generates Inner Join, in the 'memo after optimization' no right join present.
```
Comment (chrishajas) [inline src/test/regress/sql/join_gp.sql:840]: I don't understand this comment

File: src/test/regress/sql/join_gp.sql (hunk 8 above)
Comment (24nishant) [reply src/test/regress/sql/join_gp.sql:834]: I wanted to define a test case where at the time of execution we get to know information for partition elimination.
So in this query, and the query following it, I tried using "IN" clause.  

But after thinking for a while on it, please correct me if I am wrong:
Does it even make sense to have a join condition where we are not referring to both the tables, like I have done in both the test cases ? 
Normally we should have some operator between join columns of both tables. In these two query we are only referring to "bar_PT1.a1_PC in (something)". Does such condition even qualify for a valid join condition ?

File: src/test/regress/sql/join_gp.sql (hunk 9 above)
Comment (24nishant) [reply src/test/regress/sql/join_gp.sql:840]: I just wanted to highlight that, no 'Right Join' is produced in the plan, even though the query was for right join. 
Also, in this case there was no logical right join present in the MEMO itself. Thus it is not that during implementation phase the right join was removed, it was not present altogether in the plan.

File: src/backend/gporca/libgpopt/include/gpopt/operators/CPhysical.h [hunk 10]
```diff
@@ -283,6 +283,19 @@ class CPhysical : public COperator
 		CReqdPropPlan *prppInput, ULONG child_index,
 		CDrvdPropArray *pdrgpdpCtxt, ULONG ulOptReq);
 
+	CPartitionPropagationSpec *PppsRequiredForJoins(
```
Comment (chrishajas) [inline src/backend/gporca/libgpopt/include/gpopt/operators/CPhysical.h:286]: Can we put these methods in CPhysicalJoin, or CPhysicalHashJoin might be even better?

File: src/backend/gporca/libgpopt/src/operators/CPhysical.cpp [hunk 11]
```diff
@@ -1035,4 +1036,162 @@ CPhysical::PppsDerive(CMemoryPool *mp, CExpressionHandle &exprhdl) const
...
+{
+	GPOS_ASSERT(nullptr != pppsRequired);
+	GPOS_ASSERT(nullptr != pdrgpdpCtxt);
+
+	CExpression *pexprScalar = exprhdl.PexprScalarExactChild(2 /*child_index*/);
+
+	// CColRefSet *pcrsOutputOuter = exprhdl.DeriveOutputColumns(0);
+	CColRefSet *pcrsOutputInner = exprhdl.DeriveOutputColumns(1);
+
+	// CPartInfo *part_info_outer = exprhdl.DerivePartitionInfo(0);
```
Comment (chrishajas) [inline src/backend/gporca/libgpopt/src/operators/CPhysical.cpp:1078]: We can remove the commented out code here

File: src/backend/gporca/data/dxl/minidump/RightJoinDPE.mdp [hunk 12]
```diff
@@ -0,0 +1,1900 @@
...
+         insert into bar_PT1 select i,i from generate_series(1,49)i;
+         analyze bar_PT1;
+         explain select * from bar_PT1 right join foo on bar_PT1.a_PC =foo.a;
+                                                 QUERY PLAN
+        --------------------------------------------------------------------------------------------
+         Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..862.00 rows=8 width=16)
+           ->  Hash Right Join  (cost=0.00..862.00 rows=3 width=16)
+                 Hash Cond: (bar_pt1.a_pc = foo.a)
+                 ->  Dynamic Seq Scan on bar_pt1  (cost=0.00..431.00 rows=2 width=8)
+                       Number of partitions to scan: 6 (out of 6)
```
Comment (chrishajas) [inline src/backend/gporca/data/dxl/minidump/RightJoinDPE.mdp:18]: Is this the right plan for the DDL? It seems like it should have 25 partitions total

File: src/test/regress/sql/join_gp.sql (hunk 8 above)
Comment (chrishajas) [reply src/test/regress/sql/join_gp.sql:834]: It produces a hash join, but not an outer join. So while we get a partition selector, it's not due to what we added in this code

File: src/test/regress/sql/join_gp.sql [hunk 13]
```diff
@@ -768,3 +768,97 @@ full join ( select r.id1, r.id2 from t_issue_10315 r group by r.id1, r.id2 ) tq_
...
+
+-- Case 3 : Other Cases
+
+explain (costs off) select * from bar_PT1 right join foo on bar_PT1.a1_PC in (select foo.a from foo where foo.a<3);
+select * from bar_PT1 right join foo on bar_PT1.a1_PC in (select foo.a from foo where foo.a<3);
+
+explain (costs off) select * from bar_PT1 right join foo on bar_PT1.a1_PC in (select foo.a from foo where foo.a=bar_PT1.a1_PC);
+select * from bar_PT1 right join foo on bar_PT1.a1_PC in (select foo.a from foo where foo.a=bar_PT1.a1_PC);
+
+-- Scenario- Right join with predicate on partition column in 'where clause'.
```
Comment (chrishajas) [inline src/test/regress/sql/join_gp.sql:840]: The reason for this is that the predicate on the inner relation bar_PT1, is being filtered for nulls, so we can optimize this to a simple inner join. So this isn't testing the right join logic

File: src/test/regress/sql/join_gp.sql (hunk 9 above)
Comment (chrishajas) [reply src/test/regress/sql/join_gp.sql]: I see, it was optimized out. See `PexprOuterJoinToInnerJoin`. We only generate right joins from left joins. Even if the user specifies right join, we convert it a left join in Orca (see `PexprRightOuterJoin`), do all the simplifications/preprocessing rewrites, and then we have an xform that adds a right join alternative to explore when enumerating different alternatives.

File: src/test/regress/sql/join_gp.sql [hunk 14]
```diff
@@ -768,3 +768,97 @@ full join ( select r.id1, r.id2 from t_issue_10315 r group by r.id1, r.id2 ) tq_
...
+drop table if exists bar_PT1;
+drop table if exists bar_PT2;
+drop table if exists bar_PT3;
+
+-- Table creation : Normal table
+create table foo (a int, b int);
+insert into foo select i,i from generate_series(1,5)i;
+analyze foo;
+-- Table creation : First Partitioned table with same 'Distribution Column' and 'Partitioning key'
+create table bar_PT1 (a1_PC int, b1 int) partition by range(a1_PC) (start (1) inclusive end (12) every (2));
```
Comment (pobbatihari) [inline src/test/regress/sql/join_gp.sql:796]: Can you explicitly specify distribution columns while creating table.

File: src/test/regress/sql/join_gp.sql [hunk 15]
```diff
@@ -768,3 +768,97 @@ full join ( select r.id1, r.id2 from t_issue_10315 r group by r.id1, r.id2 ) tq_
...
+insert into bar_PT1 select i,i from generate_series(1,11)i;
+analyze bar_PT1;
+-- Table creation : Second Partitioned table with different 'Distribution Column' and 'Partitioning key'
+create table bar_PT2 (a2 int, b2_PC int) partition by range(b2_PC) (start (1) inclusive end (12) every (2));
```
Comment (pobbatihari) [inline src/test/regress/sql/join_gp.sql:800]: Can you explicitly specify distribution columns while creating table.

File: src/test/regress/sql/join_gp.sql [hunk 16]
```diff
@@ -768,3 +768,97 @@ full join ( select r.id1, r.id2 from t_issue_10315 r group by r.id1, r.id2 ) tq_
...
+insert into bar_PT2 select i,i from generate_series(1,11)i;
+analyze bar_PT2;
+-- Table creation : Third Partitioned table with same 'Distribution Column' and 'Partitioning key'
+create table bar_PT3 (a3_PC int, b3 int) partition by range(a3_PC) (start (1) inclusive end (6) every (2));
```
Comment (pobbatihari) [inline src/test/regress/sql/join_gp.sql:804]: Can you explicitly specify distribution columns while creating table.

File: src/backend/gporca/libgpopt/src/base/CPartitionPropagationSpec.cpp [hunk 17]
```diff
@@ -205,19 +221,19 @@ CPartitionPropagationSpec::InsertAllowedConsumers(
 		SPartPropSpecInfo *other_info =
 			const_cast<SPartPropSpecInfo *>(hmulpi.Value());
 
-		// only process allowed_scan_ids ...
```
Comment (pobbatihari) [inline src/backend/gporca/libgpopt/src/base/CPartitionPropagationSpec.cpp:208]: any reason for removing these comments

File: src/backend/gporca/data/dxl/minidump/RightJoinDPE.mdp (hunk 12 above)
Comment (24nishant) [reply src/backend/gporca/data/dxl/minidump/RightJoinDPE.mdp:18]: Yes, you are right Chris, it should have 25 partitions. I had reduced the number of partitions for ICW test, I used the plan from that here.
Now I have updated the plans.

File: src/test/regress/sql/join_gp.sql (hunk 15 above)
Comment (24nishant) [reply src/test/regress/sql/join_gp.sql:800]: Yes, I have updated in the code.

File: src/test/regress/sql/join_gp.sql (hunk 16 above)
Comment (24nishant) [reply src/test/regress/sql/join_gp.sql:804]: Yes, I have updated in the code.

File: src/test/regress/sql/join_gp.sql (hunk 9 above)
Comment (24nishant) [reply src/test/regress/sql/join_gp.sql]: Thanks @chrishajas for such insight, it really helps.

File: src/backend/gporca/libgpopt/src/operators/CPhysical.cpp [hunk 18]
```diff
@@ -26,6 +26,7 @@
 #include "gpopt/base/CReqdPropPlan.h"
 #include "gpopt/operators/CExpression.h"
 #include "gpopt/operators/CExpressionHandle.h"
+#include "gpopt/operators/CPredicateUtils.h"
```
Comment (chrishajas) [inline src/backend/gporca/libgpopt/src/operators/CPhysical.cpp:29]: Do we still need this?

**Review by chrishajas — APPROVED** [review]
Looks good!

**Review by gpopt — COMMENTED** [review]
Will CPhysicalLeftAntiSemiHashJoin or similar operators be affected by this change?

File: src/backend/gporca/libgpopt/src/operators/CPhysicalRightOuterHashJoin.cpp [hunk 19]
```diff
@@ -120,4 +121,27 @@ CPhysicalRightOuterHashJoin::CreateOptRequests(CMemoryPool *mp)
...
+
+// In the following function, we are generating the Derived property :
+// "Partition Propagation Spec" of Right Outer Hash join.
+CPartitionPropagationSpec *
+CPhysicalRightOuterHashJoin::PppsDerive(CMemoryPool *mp,
+										CExpressionHandle &exprhdl) const
+{
+	CPartitionPropagationSpec *pps_result;
+	pps_result = PppsDeriveForJoins(mp, exprhdl);
+	return pps_result;
```
Comment (gpopt) [inline src/backend/gporca/libgpopt/src/operators/CPhysicalRightOuterHashJoin.cpp:145]: Change to 
```c
return PppsDeriveForJoins(mp, exprhdl);
```

File: src/backend/gporca/libgpopt/src/operators/CPhysicalRightOuterHashJoin.cpp [hunk 20]
```diff
@@ -120,4 +121,27 @@ CPhysicalRightOuterHashJoin::CreateOptRequests(CMemoryPool *mp)
...
+CPartitionPropagationSpec *
+CPhysicalRightOuterHashJoin::PppsRequired(
+	CMemoryPool *mp, CExpressionHandle &exprhdl,
+	CPartitionPropagationSpec *pppsRequired, ULONG child_index,
+	CDrvdPropArray *pdrgpdpCtxt, ULONG ulOptReq) const
+{
+	CPartitionPropagationSpec *pps_result;
+	pps_result = PppsRequiredForJoins(mp, exprhdl, pppsRequired, child_index,
+									  pdrgpdpCtxt, ulOptReq);
+	return pps_result;
```
Comment (gpopt) [inline src/backend/gporca/libgpopt/src/operators/CPhysicalRightOuterHashJoin.cpp:134]: Also refactor this part, if the variable is not used at all, just return with function call.

File: src/backend/gporca/libgpopt/src/operators/CPhysicalInnerHashJoin.cpp [hunk 21]
```diff
@@ -292,92 +267,18 @@ CPhysicalInnerHashJoin::PppsRequired(CMemoryPool *mp,
...
-	CPartitionPropagationSpec *pps_inner = exprhdl.Pdpplan(1)->Ppps();
-
-	CPartitionPropagationSpec *pps_result =
-		GPOS_NEW(mp) CPartitionPropagationSpec(mp);
-	pps_result->InsertAll(pps_outer);
-	pps_result->InsertAllResolve(pps_inner);
-
+	CPartitionPropagationSpec *pps_result;
+	pps_result = PppsDeriveForJoins(mp, exprhdl);
 	return pps_result;
```
Comment (gpopt) [inline src/backend/gporca/libgpopt/src/operators/CPhysicalInnerHashJoin.cpp:282]: here

File: src/backend/gporca/libgpopt/src/operators/CPhysicalInnerHashJoin.cpp [hunk 22]
```diff
@@ -292,92 +267,18 @@ CPhysicalInnerHashJoin::PppsRequired(CMemoryPool *mp,
...
-	}
-	else
-	{
-		// No DPE: pass through requests
-		pps_result = CPhysical::PppsRequired(
-			mp, exprhdl, pppsRequired, child_index, pdrgpdpCtxt, ulOptReq);
-	}
+	pps_result = PppsRequiredForJoins(mp, exprhdl, pppsRequired, child_index,
+									  pdrgpdpCtxt, ulOptReq);
 	return pps_result;
```
Comment (gpopt) [inline src/backend/gporca/libgpopt/src/operators/CPhysicalInnerHashJoin.cpp:273]: here

File: src/test/regress/expected/join_gp_optimizer.out (hunk 6 above)
Comment (24nishant) [reply src/test/regress/expected/join_gp_optimizer.out:1856]: I have added test  cases for conjunction. For disjunction I tried few test cases, but in them right join in generated in the plan, thus they will not be checking the updated logic in this PR. I tried following cases 

explain select * from bar_PT1 right join foo on bar_PT1.a1_PC =foo.a or bar_PT1.a1_PC=foo.b;
explain select * from bar_PT1 right join foo on bar_PT1.a1_PC =foo.a or bar_PT1.b1=foo.b;
explain select * from bar_PT1 right join foo on bar_PT1.a1_PC =foo.a or (bar_PT1.a1_PC=foo.b and foo.a>1);

Request you to please suggest, any specific query for disjunction case.

**Review by gpopt — APPROVED** [review]
+1

File: src/test/regress/expected/join_gp_optimizer.out (hunk 6 above)
Comment (24nishant) [reply src/test/regress/expected/join_gp_optimizer.out:1906]: As discussed with @DevChattopadhyay, disjunction scenario is not helpful in DPE case, so resolving the comment.

**Review by chrishajas — APPROVED** [review]

**Pull request merged**